blueprints/hr_assistant/date.py

```diff
@@ -0,0 +1,329 @@
+        qa, size = _resolve_categorical_entities(request, responder)
+
+        # Resolve the duration that is explicitly or implicitly mentioned in the query
+        # and then filtering the KB to retain all employees satisfying that duration criteria
+        out = _resolve_time(request, responder, qa, size)
+
+        if out:
+            qa, size, field = out
+        qa_out = qa.execute(size=size)
+
+        # Calculate and return desired mathemical value
+        responder.slots['value'] = _agg_function(qa_out, func=function)
+        responder.reply('The {function} is {value}')
+
+    else:
+        responder.reply('What time-filtered statistic would you like to know?')
+        responder.listen()
+
+
+@app.handle(intent='get_date_range_employees')
+def get_date_range_employees(request, responder):
+    """
+    When a user asks for a list of employees that satisfy certain criteria in addition
+    to satisfying a specified date range criterion (date of hire/termination/birth),
+    this dialogue state filters the knowledge base on those criteria and returns the
+    shortlisted list of names.
+    """
+
+    # If there are any categorical entities (eg. race, gender, department etc.) in the query that
+    # need filtering on, '_resolve_categorical_entities' fetches these entities, resolves them to
+    # their canonical form and filters the database on all the conditions requested by the query
+    qa, size = _resolve_categorical_entities(request, responder)
+
+    # Resolve the duration that is explicitly or implicitly mentioned in the query
+    # and then filtering the KB to retain all employees satisfying that duration criteria
+    out = _resolve_time(request, responder, qa, size)
+
+    if out:
+        qa, size, field = out
+
+        # Finding extreme entities such as 'highest', 'lowest', 'youngest' etc. (if any)
+        extreme_entity = [e for e in request.entities if e['type'] == 'extreme']
+
+        # Filter on the extreme entities if they exist
+        if extreme_entity:
+            extreme_entity = extreme_entity[0]
+            qa, size = _resolve_extremes(request, responder, qa, extreme_entity, field)
+
+        qa_out = qa.execute(size=size)
+
+        if qa_out:
+            responder.slots['emp_list'] = _get_names(qa_out)
+
+            # Give the natural language response according to the size of the list to be displayed
+            if len(qa_out) == 1:
+                responder.reply("Here is the employee you are looking for: {emp_list}")
+            else:
+                responder.reply("Here are some employees that match your criteria: {emp_list}")
+
+    else:
+        responder.listen()
+
+
+# Helper functions #
+
+def _check_time_ent(time_ent, date_compare_ent):
+    """
+    Helper function for resolving non numeric time entities, time entities with
+    incompatible date formats, and intervals into the format of time that is
+    accepted by the dialogue states defined in this file.
+    """
```

Comment:
Can this be done via the numerical parser and/or entity resolver instead?

Reply:
Not always, I think. The sys_time entity only resolves to a date even when it's a year (say 2019), but the calculation for the functions here require the full interval. Hence the additional function.

Reply:
@vijay120 do you know why it is always resolving to a date? I believe Duckling gives predicted grains so it should be able to resolve years

Reply:
@ritvikshrivastava What is the problem query here?

Reply:
Bump

Reply:
@arushir So here are some queries I ran against Duckling:

Query="last week"
```
(wb3) VIJAYRK-M-L1E5:mindmeld-blueprints vijay$ curl -XPOST http://0.0.0.0:7151/parse --data 'text=last week' | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   245    0   231  100    14   127k   7900 --:--:-- --:--:-- --:--:--  225k
[
  {
    "body": "last week",
    "start": 0,
    "value": {
      "values": [
        {
          "value": "2019-08-12T00:00:00.000-07:00",
          "grain": "week",
          "type": "value"
        }
      ],
      "value": "2019-08-12T00:00:00.000-07:00",
      "grain": "week",
      "type": "value"
    },
    "end": 9,
    "dim": "time",
    "latent": false
  }
]
```

Query='last year'
(wb3) VIJAYRK-M-L1E5:mindmeld-blueprints vijay$ curl -XPOST http://0.0.0.0:7151/parse --data 'text=last year' | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   245    0   231  100    14  10646    645 --:--:-- --:--:-- --:--:-- 11000
[
  {
    "body": "last year",
    "start": 0,
    "value": {
      "values": [
        {
          "value": "2018-01-01T00:00:00.000-08:00",
          "grain": "year",
          "type": "value"
        }
      ],
      "value": "2018-01-01T00:00:00.000-08:00",
      "grain": "year",
      "type": "value"
    },
    "end": 9,
    "dim": "time",
    "latent": false
  }
]

Reply:
However, looking at the logic for this code, the intention is to get the following:
`last week`: 7 days from today
`last year`: 365 days from today

This is a different resolution result from the above.